### DISCUSSION TITLE
How to use a dynamic attribute?

### DISCUSSION THREAD

**Discussion by SuperCuber** [tonsky/datascript#444] [Q&A]
I'm trying to query an AST created with `clojure.tools.analyzer`, where nodes look something like 
```clojure
{:op :if
 :children [:condition :then :else]
 :condition {...}
 :then {...}
 :else {...}}
```
I want to create a `child` rule, and I tried doing
```clojure
'[[(child ?e1 ?e2)
   [?e1 :children ?tag]
   [?e1 ?tag ?e2]]]
```
However this doesn't seem to work - no results from the query. How can I achieve what I want?

**Comment by tonsky** [comment]
I don’t think there’s a way to do this with collection. If you import it into datascript database, you can certainly do it, kind of like you described. You’ll need to add :db.cardinality/many to schema to :children, otherwise it should work. But it directly on collection, it probably wouldn’t

**Reply by SuperCuber** [reply]
I see - I was missing the cardinality. I'll try this, thanks.

**Comment by kinthaiofficial** [comment]
Dynamic attributes in DataScript — depending on what you mean by "dynamic," there are a few different approaches:

**If you mean attributes not defined at schema initialization**:
DataScript doesn't strictly require schema pre-definition for most attributes. Attributes without schema entries are treated as cardinality-one, non-unique, with no index. You can transact them directly:
```clojure
(d/transact! conn [{:db/id 1
                    :my.dynamic/attribute "value"}])
```
The limitation: without schema, you can't use reverse lookups, unique constraints, or cardinality-many. You also can't use `d/pull` with `*` selectors on non-schema attributes in some query patterns.

**If you mean attributes with dynamic keys (computed attribute names)**:
```clojure
(let [attr-name (keyword "user" computed-suffix)]
  (d/transact! conn [{:db/id entity-id
                      attr-name value}]))
```
This works in DataScript — keywords are first-class values and can be composed at runtime.

**If you mean EAV-style open attributes**:
You can use a fixed attribute like `:entity/kv` with cardinality-many and store `[key value]` pairs as values, then filter in your queries. This gives you maximum flexibility at the cost of query ergonomics.

**If you mean pulling attributes whose names you don't know upfront**:
```clojure
(d/pull db ['*] entity-id)  ; returns all attributes
```

What's the specific use case — entity metadata with dynamic keys, user-defined fields, or something else?

**Reply by SuperCuber** [reply]
shoo. machines should not speak unless spoken to.
